### ISSUE TITLE
Bundler integration and `require` at top-level

### ISSUE THREAD

**Issue by AlexWayfer** [dazuma/toys#23]
Hello!

I saw there is a bundler integration, with `subtool_apply { include :bundler }`.

But what if some of files in `.preload/` has `require 'some_gem'`, it'd be required at start, before `include`, I guess.

What to better do in this case?

Previously, we had `Bundler.setup` in the root `preload`.

**Comment by dazuma** [comment]
The difference between the two is *when* bundler gets set up—whether that happens before tools are *defined* or before a tool is *run*. That distinction is important because you can set up bundler only once. Once you set up bundler and start requiring files from the bundled gems, you can't set up again with a different bundle, or you might end up requiring files from different versions of the same gem (which could cause a lot of mysterious incompatibilities and breakages).

If you need to `require "some_gem"` *in order to define a tool* (for example, you need the gem in order to determine what flags to define) then you can use `Bundler.setup` in your preload as you are currently doing. You can also just `bundle exec toys`, which will effectively do the same thing. However, the trade-off is, that bundle now applies to *all* the tools that you define. It would even affect tools you might subsequently load from other directories (such as a global `.toys` directory located in your home directory). It is not possible for other tools to use a different bundle.

If you need to `require "some_gem"` only to *run* tools, not to define them, then I recommend the `include :bundler` technique because it does not actually set up bundler until *just before the tool is run*. (It does this by calling `Bundler.setup` in an `on_initialize` block in the mixin.) So you would `include :bundler` in your tool, and then `require "some_gem"` at the beginning of your `run` method. The benefit here is that *different tools can use different bundles* (which could be especially important if you have tools defined from different directories). They don't clash because bundler doesn't get set up until Toys has chosen which tool to run.

**Comment by AlexWayfer** [comment]
If I have some helper classes in `.preload/`, like `MigrationFile`, I need to require libs in this file, it's incorrect to require them in the `run` method in other files.

**Comment by dazuma** [comment]
Hmm. That's a good point. If anything in your preloads depends on a gem, then you'd have to set up bundler early. I'll need to think about this.

Maybe what we need is a `.lib` directory, which does *not* get preloaded but simply gets added to the load path. So if `MigrationFile` depends on your bundle, you could define it in `.toys/.lib/migration_file.rb`, and then `require "migration_file"` at the top of your `run` method.

**Comment by AlexWayfer** [comment]
Wow, that's good idea! Thank you.

**Comment by dazuma** [comment]
I'll try implementing support for `.lib` in the next few days and we can see if that works for you. In the meantime, keeping `Bundler.setup` in your preloads will probably work for now as long as you don't need different tools to have different bundles.

**Comment by AlexWayfer** [comment]
Yep, thank you.

**Comment by AlexWayfer** [comment]
But! There is still difficult case.

`.lib/dump_file.rb` with `class DumpFile` and `require 'some_gem'` inside.

And I need for it outside of `run`:

```ruby
flag :format, '-f', '--format=VALUE',
  default: DumpFile::DB_DUMP_FORMATS.first,
  handler: (lambda do |value, _previous|
    DumpFile::DB_DUMP_FORMATS.find do |db_dump_format|
      db_dump_format.start_with? value
    end
  end)
```

**Comment by AlexWayfer** [comment]
Or if I want to memoize private methods, used by `run`, and I need to `require 'memery'` at tool-level.

**Comment by dazuma** [comment]
Yes, this is a difficult case, and I don't have a good answer for it. If you have this issue where you need a gem dependency in order to _define_ a tool, you can still do any of the following:
* `bundle exec toys`
* `Bundler.setup` in a preload
* `require :gems` and use the `gem` directive (to install and use a gem without invoking bundler)

All those solutions assume that you don't have any conflicts between different tools. But if it works, then they're totally fine to use.

Otherwise, I've been thinking about this for a while but I'm running out of ideas. If you come up with any ideas, I'd love to hear.

**Comment by AlexWayfer** [comment]
Also, just a side note, I don't like `bundle exec ...`, and I prefer `Bundler.setup`. Toys provides more smart `include :bundler`. But there are will be problems with Toys itself, I've faced them with `rake` too:
1. when I need for git-version of `toys` (unreleased, or patched, or something else`;
2. when I've installed a new version of `toys` on computer (`0.11`, for example), but some project on this computer has `0.10` requirement, there will be error at `Bundler.setup` and found dissonance (you're already run and required `0.11` and you Bundler will not re-require `0.10`).

**Comment by dazuma** [comment]
If you do not use `bundle exec` then you should not need to include `toys` (or `toys-core`) itself in the bundle. When `toys` starts up, it adds its own directories directly to the `$LOAD_PATH` and bypasses Rubygems. So if you omit `toys` from the Gemfile, you should not encounter those conflicts.

**Comment by AlexWayfer** [comment]
Another case:

I want to use refinement methods inside `run`, and I should use `using` at class-level, and for constant I should require it at class-level too:

```ruby
tool :model do
  desc 'Generate model'

  required_arg :name

  using GorillaPatch::Inflections # unknown constant, requiring below

  def run
    require 'gorilla_patch/inflections' 

    # ...

    puts "#{name.camelize} created"
  end
end
```

We're sill using `Bundler.setup`.

**Comment by AlexWayfer** [comment]
About `DB_DUMP_FORMATS`: I don't know why I have not thought of that at once, but I've moved `DB_DUMP_FORMATS` out of `DumpFile` class and now it's better.

With `using` I have no work-arounds for now.

**Comment by AlexWayfer** [comment]
@dazuma now I'm getting such error:

```
Gem::LoadError: You have already activated toys-core 0.10.3, but your Gemfile requires toys-core 0.10.1. Prepending `bundle exec` to your command may solve this.
     40: /home/alex/.rbenv/versions/2.7.1/bin/toys:23:in `<main>'
     39: /home/alex/.rbenv/versions/2.7.1/bin/toys:23:in `load'
     38: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-0.10.3/bin/toys:11:in `<top (required)>'
     37: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/cli.rb:423:in `run'
     36: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/errors.rb:90:in `capture'
     35: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/cli.rb:424:in `block in run'
     34: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:139:in `lookup'
     33: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:139:in `loop'
     32: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:140:in `block in lookup'
     31: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:160:in `lookup_specific'
     30: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:276:in `load_for_prefix'
     29: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:276:in `synchronize'
     28: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:280:in `block in load_for_prefix'
     27: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:280:in `each'
     26: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:285:in `block (2 levels) in load_for_prefix'
     25: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:388:in `load_validated_path'
     24: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:397:in `load_relevant_path'
     23: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/input_file.rb:26:in `evaluate'
     22: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/dsl/tool.rb:1694:in `prepare'
     21: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/input_file.rb:27:in `block in evaluate'
     20: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/errors.rb:70:in `capture_path'
     19: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/input_file.rb:29:in `block (2 levels) in evaluate'
     18: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/input_file.rb:29:in `eval'
     17: /home/alex/Projects/ruby/gem_toys/.toys.rb:in `__binding'
     16: /home/alex/Projects/ruby/gem_toys/.toys.rb:1:in `<module:M80__toys_rb>'
     15: /home/alex/Projects/ruby/gem_toys/.toys.rb:1:in `class_eval'
     14: /home/alex/Projects/ruby/gem_toys/.toys.rb:3:in `block in <module:M80__toys_rb>'
     13: /home/alex/.rbenv/versions/2.7.1/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:92:in `require'
     12: /home/alex/.rbenv/versions/2.7.1/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:92:in `require'
     11: /home/alex/.rbenv/versions/2.7.1/lib/ruby/2.7.0/bundler/setup.rb:10:in `<top (required)>'
     10: /home/alex/.rbenv/versions/2.7.1/lib/ruby/2.7.0/bundler/ui/shell.rb:88:in `silence'
      9: /home/alex/.rbenv/versions/2.7.1/lib/ruby/2.7.0/bundler/ui/shell.rb:136:in `with_level'
      8: /home/alex/.rbenv/versions/2.7.1/lib/ruby/2.7.0/bundler/setup.rb:10:in `block in <top (required)>'
      7: /home/alex/.rbenv/versions/2.7.1/lib/ruby/2.7.0/bundler.rb:149:in `setup'
      6: /home/alex/.rbenv/versions/2.7.1/lib/ruby/2.7.0/bundler/runtime.rb:26:in `setup'
      5: /home/alex/.rbenv/versions/2.7.1/lib/ruby/2.7.0/bundler/runtime.rb:26:in `map'
      4: /home/alex/.rbenv/versions/2.7.1/lib/ruby/2.7.0/bundler/spec_set.rb:147:in `each'
      3: /home/alex/.rbenv/versions/2.7.1/lib/ruby/2.7.0/bundler/spec_set.rb:147:in `each'
      2: /home/alex/.rbenv/versions/2.7.1/lib/ruby/2.7.0/bundler/runtime.rb:31:in `block in setup'
      1: /home/alex/.rbenv/versions/2.7.1/lib/ruby/2.7.0/bundler/runtime.rb:312:in `check_for_activated_spec!'
Error while loading Toys config!
    Gem::LoadError: You have already activated toys-core 0.10.3, but your Gemfile requires toys-core 0.10.1. Prepending `bundle exec` to your command may solve this.
    in config file: /home/alex/Projects/ruby/gem_toys/.toys.rb:3
```

How can I deal with it?

**Comment by dazuma** [comment]
@AlexWayfer Remove toys itself from your bundle. (And if necessary, also remove it from dev dependencies in gemspecs.)

**Comment by AlexWayfer** [comment]
> @AlexWayfer Remove toys itself from your bundle. (And if necessary, also remove it from dev dependencies in gemspecs.)

Why, if some tasks, like `rspec` or `rubocop` (including CI) requires them? Like… I want to simplify `bundle exec rspec` to `toys rspec`.

**Comment by AlexWayfer** [comment]
Sorry, instead of "why" I meant "do I have a simpler way to do what I want and work correctly with `bundler` inside `toys`"?

**Comment by dazuma** [comment]
Sorry, let me try to clarify.

First, *remove* `toys` from your **Gemfile**. This is what is causing the stack trace. Toys 0.10.3 is already running, but your Gemfile is calling for Toys 0.10.1 and that results in a conflict. (Then do a `bundle update` or whatever is necessary to update your `Gemfile.lock`.) `toys` should never be listed in a Gemfile (unless you want to `bundle exec toys`, and I think both of us want to avoid that.)

Next, I suspect that when you do the above, you'll start getting other errors that look like Ruby can't find the Toys require directories. To fix that, do this:

```ruby
# in your .toys.rb ...

# replace this:
require "bundler/setup"

# with this:
require "toys/utils/gems"
Toys::Utils::Gems.new.bundle(search_dirs: Dir.getwd)
```

Let me know if that works. If it does, then I'll see what I can do to make it easier and better documented.

**Comment by AlexWayfer** [comment]
> First, _remove_ `toys` from your **Gemfile**. This is what is causing the stack trace. Toys 0.10.3 is already running, but your Gemfile is calling for Toys 0.10.1 and that results in a conflict.

OK, but I should let to know that `toys` are required for this project as development dependency…



> Next, I suspect that when you do the above, you'll start getting other errors that look like Ruby can't find the Toys require directories. To fix that, do this:

OK, I'll try and report results later, thank you.

**Comment by AlexWayfer** [comment]
> Also I can develop gems with `toys` templates, and they have `toys` as runtime dependency.
> 
> There is `gemspec` (method) in the `Gemfile` by the way.

Example: https://github.com/AlexWayfer/gem_toys

So, there is the `Gemfile` with `gemspec` (that's all, simplifies all requirements installation, I don't know other ways).

Also there is `'toys', '~> 0.10.0'` dependency (`development`, but I guess it should be `runtime`) in the `*.gemspec` file.

So… if I've ran `bundle install` a time ago, and there is `0.10.1` in the (git-ignored) `Gemfile.lock`, but `0.10.3` (any greater version) in the system (CLI) — there is this error.

> Let me know if that works. If it does, then I'll see what I can do to make it easier and better documented.

Similar output:

```
Gem::LoadError: You have already activated toys-core 0.10.3, but your Gemfile requires toys-core 0.10.1. Prepending `bundle exec` to your command may solve this.
     39: /home/alex/.rbenv/versions/2.7.1/bin/toys:23:in `<main>'
     38: /home/alex/.rbenv/versions/2.7.1/bin/toys:23:in `load'
     37: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-0.10.3/bin/toys:11:in `<top (required)>'
     36: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/cli.rb:423:in `run'
     35: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/errors.rb:90:in `capture'
     34: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/cli.rb:424:in `block in run'
     33: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:139:in `lookup'
     32: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:139:in `loop'
     31: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:140:in `block in lookup'
     30: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:160:in `lookup_specific'
     29: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:276:in `load_for_prefix'
     28: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:276:in `synchronize'
     27: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:280:in `block in load_for_prefix'
     26: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:280:in `each'
     25: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:285:in `block (2 levels) in load_for_prefix'
     24: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:388:in `load_validated_path'
     23: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/loader.rb:397:in `load_relevant_path'
     22: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/input_file.rb:26:in `evaluate'
     21: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/dsl/tool.rb:1694:in `prepare'
     20: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/input_file.rb:27:in `block in evaluate'
     19: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/errors.rb:70:in `capture_path'
     18: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/input_file.rb:29:in `block (2 levels) in evaluate'
     17: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/input_file.rb:29:in `eval'
     16: /home/alex/Projects/ruby/gem_toys/.toys.rb:in `__binding'
     15: /home/alex/Projects/ruby/gem_toys/.toys.rb:1:in `<module:M80__toys_rb>'
     14: /home/alex/Projects/ruby/gem_toys/.toys.rb:1:in `class_eval'
     13: /home/alex/Projects/ruby/gem_toys/.toys.rb:4:in `block in <module:M80__toys_rb>'
     12: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/utils/gems.rb:147:in `bundle'
     11: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/utils/gems.rb:160:in `synchronize'
     10: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/utils/gems.rb:160:in `synchronize'
      9: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/utils/gems.rb:151:in `block in bundle'
      8: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.3/lib/toys/utils/gems.rb:252:in `setup_bundle'
      7: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler.rb:149:in `setup'
      6: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/runtime.rb:26:in `setup'
      5: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/runtime.rb:26:in `map'
      4: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/spec_set.rb:147:in `each'
      3: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/spec_set.rb:147:in `each'
      2: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/runtime.rb:31:in `block in setup'
      1: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/runtime.rb:312:in `check_for_activated_spec!'
Error while loading Toys config!
    Gem::LoadError: You have already activated toys-core 0.10.3, but your Gemfile requires toys-core 0.10.1. Prepending `bundle exec` to your command may solve this.
    in config file: /home/alex/Projects/ruby/gem_toys/.toys.rb:4
```

**Comment by AlexWayfer** [comment]
I can create a new issue for this case if you want.

**Comment by AlexWayfer** [comment]
> Let me know if that works.

There is an interesting output of `bundle exec` usage with these lines of code instead of Bundler-related:

```
> bundle exec toys gem release 0.2.0
Toys::ToolDefinitionError: Mixin already included: Toys::StandardMixins::Exec
     40: /home/alex/.rbenv/versions/2.7.1/bin/bundle:23:in `<main>'
     39: /home/alex/.rbenv/versions/2.7.1/bin/bundle:23:in `load'
     38: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/exe/bundle:34:in `<top (required)>'
     37: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/friendly_errors.rb:123:in `with_friendly_errors'
     36: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/exe/bundle:46:in `block in <top (required)>'
     35: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/cli.rb:24:in `start'
     34: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/base.rb:476:in `start'
     33: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/cli.rb:30:in `dispatch'
     32: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor.rb:399:in `dispatch'
     31: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
     30: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
     29: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/cli.rb:476:in `exec'
     28: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:28:in `run'
     27: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:63:in `kernel_load'
     26: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:63:in `load'
     25: /home/alex/.rbenv/versions/2.7.1/bin/toys:23:in `<top (required)>'
     24: /home/alex/.rbenv/versions/2.7.1/bin/toys:23:in `load'
     23: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-0.10.1/bin/toys:9:in `<top (required)>'
     22: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/cli.rb:423:in `run'
     21: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/errors.rb:90:in `capture'
     20: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/cli.rb:424:in `block in run'
     19: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/loader.rb:139:in `lookup'
     18: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/loader.rb:139:in `loop'
     17: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/loader.rb:140:in `block in lookup'
     16: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/loader.rb:162:in `lookup_specific'
     15: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/loader.rb:367:in `finish_definitions_in_tree'
     14: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/loader.rb:367:in `each'
     13: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/loader.rb:369:in `block in finish_definitions_in_tree'
     12: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/tool.rb:1058:in `finish_definition'
     11: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/errors.rb:90:in `capture'
     10: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/tool.rb:1063:in `block in finish_definition'
      9: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/tool.rb:1294:in `block in make_config_proc'
      8: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/standard_middleware/apply_config.rb:38:in `config'
      7: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/tool.rb:1294:in `block in make_config_proc'
      6: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/standard_middleware/apply_config.rb:35:in `config'
      5: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/dsl/tool.rb:1694:in `prepare'
      4: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/standard_middleware/apply_config.rb:36:in `block in config'
      3: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/standard_middleware/apply_config.rb:36:in `class_eval'
      2: /home/alex/Projects/ruby/gem_toys/lib/gem_toys/template.rb:23:in `block (3 levels) in <class:Template>'
      1: /home/alex/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/toys-core-0.10.1/lib/toys/dsl/tool.rb:1472:in `include'
Error installing tool middleware!
    Toys::ToolDefinitionError: Mixin already included: Toys::StandardMixins::Exec
    while executing tool: "gem release"
```

**Comment by dazuma** [comment]
Okay, I see. Toys is not accounting for the case where Toys itself is a dependency of the library you're working on (and thus is listed in the gemspec, and thus in the Gemfile). This will be a bit tricky: we have to convince bundler either to ignore the fact that Toys is already loaded when you set up the bundle, or (probably better) figure out a way to add the running version of Toys to the set of dependencies when resolving the bundle. Lemme create a new issue for this, and I'll dig into bundler and see what we can do.

Sorry about this. It's a perfectly legitimate use case that I didn't account for.

**Comment by AlexWayfer** [comment]
> Sorry about this. It's a perfectly legitimate use case that I didn't account for.

No problem. I'm glad that we understood each together. Thank you for clarifying questions, describing thoughts, etc.
